kaldb/src/main/java/com/slack/kaldb/logstore/schema/SchemaAwareLogDocumentBuilderImpl.java

```diff
@@ -0,0 +1,344 @@
+package com.slack.kaldb.logstore.schema;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.slack.kaldb.logstore.DocumentBuilder;
+import com.slack.kaldb.logstore.FieldDefMismatchException;
+import com.slack.kaldb.logstore.LogMessage;
+import com.slack.kaldb.metadata.schema.FieldType;
+import com.slack.kaldb.metadata.schema.LuceneFieldDef;
+import com.slack.kaldb.util.JsonUtil;
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.logging.log4j.util.Strings;
+import org.apache.lucene.document.Document;
+import org.apache.lucene.document.Field;
+import org.apache.lucene.document.StoredField;
+import org.apache.lucene.document.StringField;
+import org.apache.lucene.document.TextField;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * SchemaAwareLogDocumentBuilder always indexes a field using the same type. It doesn't allow field
+ * conflicts.
+ *
+ * <p>In case of a field conflict, this class uses FieldConflictPolicy to handle them.
+ *
+ * <p>NOTE: Currently, if building a document raises errors, we still register the type of the
+ * fields in this document partially. While the document may not be indexed, this partial field
+ * config will exist in the system. For now, we assume storing this metadata is fine since it is
+ * rarely an issue and helps with performance. If this is an issue, we need to scan the json twice
+ * to ensure document is good to index.
+ */
+public class SchemaAwareLogDocumentBuilderImpl implements DocumentBuilder<LogMessage> {
```

Comment:
can we not change the package name in this PR? Or let's do it in a previous PR?

It's impossible to tell what changed in this class since the package was moved and also the number of lines here is much lesser.

So to make it a little more friendly to review can we wither list down the changes from the previous impl, or do the package change in a separate PR?

Reply:
Done.